src/BinSkim.Sdk/Telemetry.cs

```diff
@@ -0,0 +1,166 @@
+
+        /// <summary>
+        /// Gets the telemetry configuration. May be null if Application Insights
+        /// telemetry is not enabled.
+        /// </summary>
+        private TelemetryConfiguration? TelemetryConfiguration { get; }
+
+        /// <summary>
+        /// Gets the telemetry client. This may be null if Application Insights
+        /// telemetry is not enabled.
+        /// </summary>
+        public TelemetryClient? TelemetryClient { get; }
+
+        /// <summary>
+        /// Construct a new <see cref="Telemetry"/> instance from environment variables.
+        /// </summary>
+        /// <remarks>
+        /// If environment variables aren't set, then <see cref="TelemetryClient"/> will be null.
+        /// </remarks>
+        public Telemetry() : this(CreateTelemetryConfigurationFromEnvironment())
+        {
+        }
+
+        /// <summary>
+        /// Construct a new <see cref="Telemetry"/> instance using
+        /// the given <see cref="ApplicationInsights.Extensibility.TelemetryConfiguration"/>.
+        /// </summary>
+        /// <param name="telemetryConfiguration">The pipeline to use. May be null, in which case
+        /// Application Insights is disabled.</param>
+        /// <remarks>
+        /// If <paramref name="telemetryConfiguration"/> is non-null, it will be
+        /// disposed when this object is disposed.
+        /// </remarks>
+        internal Telemetry(TelemetryConfiguration? telemetryConfiguration)
+        {
+            if (telemetryConfiguration != null)
+            {
+                var telemetryClient = new TelemetryClient(telemetryConfiguration);
+                ConfigureTelemetryContext(telemetryClient.Context);
+                this.TelemetryClient = telemetryClient;
+                this.TelemetryConfiguration = telemetryConfiguration;
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            // Flush and close AppInsights client.
+            this.TelemetryClient?.FlushAsync(CancellationToken.None).GetAwaiter().GetResult();
+            this.TelemetryConfiguration?.Dispose();
+        }
+
+        private static void ConfigureTelemetryContext(TelemetryContext context)
+        {
+            context.Session.Id = CreateRandomSessionId();
+            context.Component.Version = Assembly.GetCallingAssembly().GetCustomAttribute<AssemblyFileVersionAttribute>().Version;
+            context.Device.OperatingSystem = RuntimeInformation.OSDescription;
+        }
+
+        private static TelemetryConfiguration? CreateTelemetryConfigurationFromEnvironment()
+        {
+            string? appInsightsConnectionString = CompilerDataLogger.RetrieveEnvironmentVariable(AppInsightsConnectionStringEnvVar);
+            if (!string.IsNullOrEmpty(appInsightsConnectionString))
+            {
+                return new TelemetryConfiguration
```

Comment:
>TelemetryConfiguration [](http://example.com/codeflow?start=27&length=22)

What happens if we pass a bad key here? Exception? Did you observe behavior for this case? What should our behavior be? On the one hand, silently failing to generate telemetry would be bad. On the other, anyone setting a bad key could bring down every binskim process that subsequently runs.

I think I'd favor failing immediately, ensuring sufficiently informative exception details on the console.

Thoughts?

Reply:
I'll check and add a unit test, if necessary.

Reply:
There is validation in the ConnectionString setter. It throws an ArgumentException if it can't be parsed as "key=value" pairs with semicolon separators and also if InstrumentationKey is missing. I'm happy with that. It'll fail immediately with a meaningful exception.